run `gvproxy` with its services API enabled

anjannath opened the pull request:
if we need to add DNS entries or port forwarding for a VM
then the gvproxy process provides an API to add those dns
records or port forwards, but currently it is ran without
a listener for the rest API

this change adds a helper to get a path in the VM runtime
dir for a unix socket which can be used to expose the API

anjannath committed:
vmconfig: add helper to get gvproxy services socket path

if we need to add DNS entries or port forwarding for a VM
then the gvproxy process provides an API to add those dns
records or port forwards, but currently it is ran without
a listener for the rest API

this change adds a helper to get a path in the VM runtime
dir for a unix socket which can be used to expose the API

cfergeau commented on pkg/machine/vmconfigs/sockets.go:
```diff
@@ -19,3 +19,8 @@ func readySocket(name string, machineRuntimeDir *define.VMFile) (*define.VMFile,
+
+func gvProxyServicesSocket(name string, machineRuntimeDir *define.VMFile) (*define.VMFile, error) {
+	endpoint := fmt.Sprintf("%s-gvproxy-api.sock", name)
+	return machineRuntimeDir.AppendToNewVMFile(endpoint, &endpoint)
```
On `!darwin`, the last arg (`&endpoint`) can be `nil`. This arg is used to play symlink tricks on macos where unix socket paths are limited to ~100 bytes.

cfergeau reviewed (commented):
For hyper-v, I assume a similar change will be needed after usermode networking is enabled? Maybe this could already be part of these changes since podman machine is already using usermode networking on macos?

Are there any security concerns to have after this is enabled? This opens some new ways to tweak container networking.

anjannath commented:
> For hyper-v, I assume a similar change will be needed after usermode networking is enabled? Maybe this could already be part of these changes since podman machine is already using usermode networking on macos?
> 

added the changes for hyper-v as well

> Are there any security concerns to have after this is enabled? This opens some new ways to tweak container networking.

i think we need to prevent API access from inside the VM through the gateway ip if this is possible, but adding/removing entries to the DNS will only be effective for the VM, so the host side DNS won't change, and the `gvproxy` process itself is not running with higher privileges, so it cannot bind to lower ports

cfergeau commented:
> the gvproxy process itself is not running with higher privileges, so it cannot bind to lower ports

We have this in crc though, I think it’s meant to allow the daemon to bind to port 443 https://github.com/crc-org/crc/blob/11140c50b5cfa86be22a440f488cc0262d52d0dd/pkg/crc/preflight/preflight_linux.go#L224-L227 (in other words, something to keep in mind which might need fixing)

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -231,6 +232,19 @@ func (h HyperVStubber) StartNetworking(mc *vmconfigs.MachineConfig, cmd *gvproxy
 		cmd.AddEndpoint(fmt.Sprintf("vsock://%s", mc.HyperVHypervisor.NetworkVSock.KeyName))
 		return nil
```
The new code should go in this block

anjannath replied:
ahh, yes

cfergeau commented on pkg/machine/apple/apple.go:
```diff
@@ -445,6 +445,20 @@ func StartGenericNetworking(mc *vmconfigs.MachineConfig, cmd *gvproxy.GvproxyCom
+	if err != nil {
+		return err
+	}
+	cmd.AddServiceEndpoint(serviceSocket.String())
```
kind of annoyed to see this code duplicated 3 times, would there be a good place where it could be shared?

anjannath replied:
there's `pkg/machine/gvproxy.go`, we could add a helper there: 
```go
func AddGvproxyServiceCmdline(mc *vmconfigs.MachineConfig, cmd *gvproxy.GvproxyCommand) (*gvproxy.GvproxyCommand, error)
```

cfergeau replied:
Could this go in `shim/networking.go` in `startHostForwarder`?

anjannath replied:
thanks, updated!

anjannath commented:
> > the gvproxy process itself is not running with higher privileges, so it cannot bind to lower ports
> 
> We have this in crc though, I think it’s meant to allow the daemon to bind to port 443 https://github.com/crc-org/crc/blob/11140c50b5cfa86be22a440f488cc0262d52d0dd/pkg/crc/preflight/preflight_linux.go#L224-L227 (in other words, something to keep in mind which might need fixing)

right, bot port 80 and 443 for the ocp web console routes, can we release the binary with the capability already added, or the `setcap` needs to be ran on the user machine for it to work?

cfergeau commented:
> > > the gvproxy process itself is not running with higher privileges, so it cannot bind to lower ports
> > 
> > 
> > We have this in crc though, I think it’s meant to allow the daemon to bind to port 443 https://github.com/crc-org/crc/blob/11140c50b5cfa86be22a440f488cc0262d52d0dd/pkg/crc/preflight/preflight_linux.go#L224-L227 (in other words, something to keep in mind which might need fixing)
> 
> right, bot port 80 and 443 for the ocp web console routes, can we release the binary with the capability already added, or the `setcap` needs to be ran on the user machine for it to work?

The RPM package could do this, not sure we want to always have this. The binary on github cannot have the capability added automatically when it’s downloaded.

anjannath committed:
networking: start gvproxy with services API

this adds the '-services' flag to the 'gvproxy'
commandline with a unix socket so the REST API
for managing DNS and port forward is available
